Issue #4860 - NPE from HttpFields

gregw opened the pull request:
Fix handling of null fields in `HttpFields` for #4860 :
 + Added paranoid catch in `HttpChannel` so even if fix doesn't work we won't loop forever.
 + Fixed bug in iterator with nextIndex
 + Do not allow null fields to be set with list iterator
 + improved test coverage

gregw committed:
Issue #4860 NPE from HttpFields

Paranoid catch if sending and exception page throws an exception.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4860 NPE from HttpFields

Improved test coverage

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4860 NPE from HttpFields

 + Fix bug with list iterator nextIndex
 + List iterator cannot inject null fields
 + minor cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet and @lachlan-roberts I think this issue/fix could relate to our previous discussions about if httpFields.add(null) should be a noop or throw.   If this problem is indeed caused by springboot doing a `httpFields.listIterator().next().set(null)`, then we do have users that are expecting the noop.... but then we are breaking the api in 10...

joakime reviewed (commented):
I generally like what I'm seeing.

Just wondering if we need to beef up the javadoc / apidoc for this class too?

gregw commented:
@joakime This class is substantially rewritten in 10, so I think javadoc effort should go there.   Will be interesting to take new tests to 10 as well.

I'm still not 100% convinced this is the actual problem, but can't see another way for a null to get in the array.

joakime commented:
The OP did say this was during a SSE behavior, so there could be access to the response object after the connection is closed / recycled, no?

joakime committed:
Eliminate warnings

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #4860 - Improving testPreventNullField testcase

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #4860 - Consistency to HttpFields API

If a null name (or HttpHeader or HttpField) is used
it should throw an ISE

+ .add() should remain consistent
+ .put() should remain consistent

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
@gregw the updated testing and changes to HttpFields for null name has been committed to this branch.

gregw commented on jetty-http/src/test/java/org/eclipse/jetty/http/HttpFieldsTest.java:
```diff
@@ -710,15 +749,15 @@ public void testPreventNullField()
         ListIterator<HttpField> iter = fields.listIterator();
         iter.next();
-        iter.set(null);
+        iter.set(null); // set field to null - null entry in list now
```
@joakime This comment is wrong.  Setting null used to put in a null entry, but now is a noop.

gregw commented:
@sbordet Should we change code like:
```
_fields[_size++] = field;
```
to
```
_fields[_size] = field;
_size++;
```

This class is not required to be Thread safe, but doing the updates in that order (and the reverse for deletion) might help avoid NPEs if bad apps are concurrently mutating.... or is the jit just going to reorder anyway, so no point?

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpField.java:
```diff
@@ -40,7 +40,7 @@ public HttpField(HttpHeader header, String name, String value)
             _name = _header.asString();
         else
-            _name = Objects.requireNonNull(name);
+            _name = Objects.requireNonNull(name, "name");
```
Since you bothered adding a message to the NPE, make it "header name cannot be null", or at least "header name", as "name" alone is just too laconic to be useful.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -671,6 +676,8 @@ public void put(HttpHeader header, HttpHeaderValue value)
      */
     public void put(HttpHeader header, String value)
     {
+        Objects.requireNonNull(header, "header");
```
"header cannot be null".

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -685,7 +692,12 @@ public void put(HttpHeader header, String value)
      */
     public void put(String name, List<String> list)
     {
+        Objects.requireNonNull(name, "name");
```
"name cannot be null"

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -723,6 +736,8 @@ public void add(HttpHeader header, HttpHeaderValue value)
      */
     public void add(HttpHeader header, String value)
     {
+        Objects.requireNonNull(header, "header");
```
"header cannot be null"

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -723,6 +736,8 @@ public void add(HttpHeader header, HttpHeaderValue value)
+        Objects.requireNonNull(header, "header");
+
         if (value == null)
             throw new IllegalArgumentException("null value");
```
We are no-op all the other cases, most notably `add(String, String)` above, so why throw here?

gregw replied:
Because it will have servlet API impacts.  There is a separate issue that I thought you and and @lachlan-roberts  were working on to normalise the API in 10 to either always noop or always ISE (I can't remember which).  Either way, I don't want to change behaviour in this PR other than to avoid noops and later NPEs (eg from null names or null fields).

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -1199,16 +1217,22 @@ public void set(HttpField field)
+            if (field != null)
+            {
+                _fields = Arrays.copyOf(_fields, _fields.length + 1);
+                System.arraycopy(_fields, _cursor, _fields, _cursor + 1, _size++);
```
Should not the last parameter be `_size++ - _cursor`?
And yes, I would rather separate the `++` and `--` to different lines so that they are more comprehensible.
For example, it takes too much brain power to know if `5++ - 3` is `2` or `3`.

We have tests that call `add(HttpField)` but the error in the copy does not show because the array is initialized to 16 elements but in the tests _size is only 4-5.

gregw replied:
Yep it's a bug and if I change the capacity to 5 in the test I get a failure... but I don't see that it is the bug we are looking for :(

gregw committed:
Issue #4860 NPE HttpFields

Fixes from review.
Fixed iterator overflow bug
clearer updates of size
better nonNull messages

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime committed:
Issue #4860 - Fixing test comment

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Merge branch 'jetty-9.4.x-4860-NullHttpFields' of github.com:eclipse/jetty.project into jetty-9.4.x-4860-NullHttpFields
